File: among_them/guided_bot/perception/geometry.nim
## Coordinate / camera math. Phase 1.2.
##
## Pure functions. Port of ``modulabot/geometry.py``. Two coordinate
## systems are in play:
##
## - **World** coordinates: pixel positions inside the 952×534 map.
##   Tasks, rooms, vents, the emergency button, and the home point all
##   live in world coordinates.
## - **Screen** coordinates: 128×128 with the player sprite's visual
##   centre at :data:`PlayerScreenX` / :data:`PlayerScreenY`. The
##   camera ``(x, y)`` is the world offset of the screen's top-left
##   pixel.
##
## The player's collision box is drawn at a small offset from the
## screen centre, so inferring world position from a locked camera is
## ``camera + PlayerWorldOff``.
##
## All procs here are pure: no module state, no allocation. They're
## hot in the localizer's local-refit and in every consumer that
## wants to convert between camera and world coords.

import ../constants
import data

const
  ## Screen-space anchor of the player sprite. The renderer centres
  ## the player and scrolls the map under them.
  PlayerScreenX* = ScreenWidth div 2
  PlayerScreenY* = ScreenHeight div 2

  ## World-space offset from the camera origin to the player's
  ## inferred collision-box centre. ``PlayerWorldOffX = SpriteDrawOffX
  ## + PlayerScreenX - SpriteSize div 2`` (and similarly for Y),
  ## verbatim from ``modulabot/geometry.py``.
  PlayerWorldOffX* =
    SpriteDrawOffX + PlayerScreenX - (SpriteSize div 2)
  PlayerWorldOffY* =
    SpriteDrawOffY + PlayerScreenY - (SpriteSize div 2)

# ---------------------------------------------------------------------------
# Camera bounds
# ---------------------------------------------------------------------------

proc minCameraX*(): int {.inline.} =
  -ScreenWidth div 2 - SpriteSize

proc maxCameraX*(): int {.inline.} =
  MapWidth - ScreenWidth div 2 + SpriteSize

proc minCameraY*(): int {.inline.} =
  -ScreenHeight div 2 - SpriteSize

proc maxCameraY*(): int {.inline.} =
  MapHeight - ScreenHeight div 2 + SpriteSize

proc cameraIndex*(x, y: int): int {.inline.} =
  ## Linear index into a (cameraH, cameraW) grid keyed on (cy, cx).
  ## Used by the localizer's vote accumulator.
  (y - minCameraY()) * (maxCameraX() - minCameraX() + 1) +
    (x - minCameraX())

proc cameraIndexX*(idx: int): int {.inline.} =
  minCameraX() + idx mod (maxCameraX() - minCameraX() + 1)

proc cameraIndexY*(idx: int): int {.inline.} =
  minCameraY() + idx div (maxCameraX() - minCameraX() + 1)

proc cameraWidth*(): int {.inline.} =
  maxCameraX() - minCameraX() + 1

proc cameraHeight*(): int {.inline.} =
  maxCameraY() - minCameraY() + 1

# ---------------------------------------------------------------------------
# Button / world helpers
# ---------------------------------------------------------------------------

proc clamp(v, lo, hi: int): int {.inline.} =
  if v < lo: lo elif v > hi: hi else: v

proc clamp(v, lo, hi: float): float {.inline.} =
  if v < lo: lo elif v > hi: hi else: v

proc buttonCameraX*(map: GameMap): int =
  ## Initial camera X guess centred on the emergency button. Clamped
  ## to the camera-range so off-edge buttons don't push the seed past
  ## the bounds.
  let target = map.button.x + map.button.w div 2 - PlayerWorldOffX
  clamp(target, minCameraX(), maxCameraX())

proc buttonCameraY*(map: GameMap): int =
  let target = map.button.y + map.button.h div 2 - PlayerWorldOffY
  clamp(target, minCameraY(), maxCameraY())

proc cameraXForWorld*(x: int): int =
  ## Camera X that puts world ``x`` on the player.
  clamp(x - PlayerWorldOffX, minCameraX(), maxCameraX())

proc cameraYForWorld*(y: int): int =
  clamp(y - PlayerWorldOffY, minCameraY(), maxCameraY())

proc inMap*(x, y: int): bool {.inline.} =
  x >= 0 and x < MapWidth and y >= 0 and y < MapHeight

proc cameraCanHoldPlayer*(cx, cy: int): bool {.inline.} =
  ## True when the camera puts the player's inferred world position
  ## inside the map rectangle. Filters out impossible candidates the
  ## patch-vote step would otherwise score.
  inMap(cx + PlayerWorldOffX, cy + PlayerWorldOffY)

# ---------------------------------------------------------------------------
# Player-from-camera helpers
# ---------------------------------------------------------------------------

proc playerWorldX*(cameraX: int): int {.inline.} =
  cameraX + PlayerWorldOffX

proc playerWorldY*(cameraY: int): int {.inline.} =
  cameraY + PlayerWorldOffY

# ---------------------------------------------------------------------------
# Room lookups
# ---------------------------------------------------------------------------

proc roomNameAt*(map: GameMap, x, y: int): string =
  ## Linear scan over the rooms list — there are ~30 rooms; not worth
  ## a spatial index. Returns ``"unknown"`` for points outside every
  ## room.
  for room in map.rooms:
    if x >= room.x and x < room.x + room.w and
       y >= room.y and y < room.y + room.h:
      return room.name
  "unknown"

proc visibleCrewmateWorldX*(cameraX, screenX: int): int {.inline.} =
  cameraX + screenX + SpriteDrawOffX

proc visibleCrewmateWorldY*(cameraY, screenY: int): int {.inline.} =
  cameraY + screenY + SpriteDrawOffY

proc heuristic*(ax, ay, bx, by: int): int {.inline.} =
  ## Manhattan distance — same name as in modulabot/geometry.py to
  ## keep consumers (A\*, sortBy distance) symmetrical.
  abs(ax - bx) + abs(ay - by)

# ---------------------------------------------------------------------------
# Task-station helpers (phase 6.1)
# ---------------------------------------------------------------------------

proc taskIconExpectedScreenPos*(station: TaskStation,
                                camX, camY: int): (int, int) =
  ## Expected screen position of the task icon for a station, given
  ## the current camera. The icon renders above the station rect
  ## centre. Mirrors ``modulabot/perception/pixel_pipeline.py``'s
  ## ``_task_icon_expected_pos``.
  let wx = station.x + station.w div 2
  let wy = station.y + station.h div 2
  let sx = wx - camX - SpriteDrawOffX
  let sy = wy - camY - SpriteDrawOffY
  (sx, sy)

proc taskIconOnScreen*(station: TaskStation,
                       camX, camY: int,
                       margin: int): bool =
  ## True when the task icon's expected position is fully inside the
  ## screen bounds with ``margin`` pixels of clearance on each edge.
  ## Used to gate icon-miss counting: we only count a miss when we're
  ## confident the icon *would* be visible if it existed.
  let (sx, sy) = taskIconExpectedScreenPos(station, camX, camY)
  # The icon sprite is SpriteSize x SpriteSize. Check that the full
  # sprite rect (anchored at sx, sy) is within the margin-inset screen.
  sx >= margin and
  sy >= margin and
  sx + SpriteSize <= ScreenWidth - margin and
  sy + SpriteSize <= ScreenHeight - margin

proc projectedRadarDot*(station: TaskStation,
                        camX, camY: int,
                        playerWx, playerWy: int): (int, int) =
  ## Where the server would draw a radar dot for this off-screen station.
  ## Uses ray-clip from player screen position to icon screen position,
  ## matching the server's projection algorithm faithfully.
  ## See pixel_pipeline.py:258-320 for the Python reference port.
  let iconSx = station.x + station.w div 2 - SpriteSize div 2 - camX
  let iconSy = station.y - SpriteSize - 2 - camY
  let iconX = iconSx + SpriteSize div 2
  let iconY = iconSy + SpriteSize div 2

  # If icon is on-screen, the server draws the icon, not a dot.
  # Caller should check taskIconOnScreen() first; this returns the
  # icon centre for convenience but it's not a dot position.
  if iconSx + SpriteSize > 0 and iconSy + SpriteSize > 0 and
     iconSx < ScreenWidth and iconSy < ScreenHeight:
    return (iconX, iconY)

  let px = float(playerWx - camX)
  let py = float(playerWy - camY)
  let dx = float(iconX) - px
  let dy = float(iconY) - py

  if abs(dx) < 0.5 and abs(dy) < 0.5:
    return (0, 0)

  var ex, ey: float
  if abs(dx) > abs(dy):
    ex = if dx > 0.0: float(ScreenWidth - 1) else: 0.0
    ey = py + dy * (ex - px) / dx
    ey = clamp(ey, 0.0, float(ScreenHeight - 1))
  else:
    ey = if dy > 0.0: float(ScreenHeight - 1) else: 0.0
    ex = px + dx * (ey - py) / dy
    ex = clamp(ex, 0.0, float(ScreenWidth - 1))
  (int(ex), int(ey))

proc projectedRadarDot*(station: TaskStation,
                        camX, camY: int): (int, int) {.deprecated.} =
  ## Backward-compatible wrapper for older callers that infer player
  ## position from the camera lock.
  projectedRadarDot(station, camX, camY,
                    playerWorldX(camX), playerWorldY(camY))

File: among_them/modulabot/geometry.py
"""Coordinate and camera math.

Pure functions. Port of ``geometry.nim``. No side effects; every proc is
a simple arithmetic helper.

Two coordinate systems are in play:

- **World** coordinates: pixel positions inside the 952×534 map. This is
  what tasks, rooms, vents, the emergency button, and the ``home`` point
  live in.
- **Screen** coordinates: 128×128 with the player sprite's visual centre
  at :data:`PLAYER_SCREEN_X` / :data:`PLAYER_SCREEN_Y`. The camera X/Y
  is the world offset of the screen's top-left pixel.

The player's collision box is drawn at a small offset from the screen
centre (:data:`PLAYER_WORLD_OFF_X` / ``Y``), so inferring world-position
from a locked camera is ``camera + PLAYER_WORLD_OFF``.
"""

from __future__ import annotations

from .data import (
    MAP_HEIGHT,
    MAP_WIDTH,
    SCREEN_HEIGHT,
    SCREEN_WIDTH,
    SPRITE_DRAW_OFF_X,
    SPRITE_DRAW_OFF_Y,
    SPRITE_SIZE,
    GameMap,
    Rect,
    TaskStation,
)
from .state import Perception

PLAYER_SCREEN_X = SCREEN_WIDTH // 2
PLAYER_SCREEN_Y = SCREEN_HEIGHT // 2
PLAYER_WORLD_OFF_X = SPRITE_DRAW_OFF_X + PLAYER_SCREEN_X - SPRITE_SIZE // 2
PLAYER_WORLD_OFF_Y = SPRITE_DRAW_OFF_Y + PLAYER_SCREEN_Y - SPRITE_SIZE // 2


def min_camera_x() -> int:
    """Smallest centred camera X (half-screen + sprite overhang)."""
    return -SCREEN_WIDTH // 2 - SPRITE_SIZE


def max_camera_x() -> int:
    return MAP_WIDTH - SCREEN_WIDTH // 2 + SPRITE_SIZE


def min_camera_y() -> int:
    return -SCREEN_HEIGHT // 2 - SPRITE_SIZE


def max_camera_y() -> int:
    return MAP_HEIGHT - SCREEN_HEIGHT // 2 + SPRITE_SIZE


def camera_index(x: int, y: int) -> int:
    """Linear index into the localizer's vote/hash table for one camera offset."""
    return (y - min_camera_y()) * (max_camera_x() - min_camera_x() + 1) + (x - min_camera_x())


def camera_index_x(idx: int) -> int:
    return min_camera_x() + idx % (max_camera_x() - min_camera_x() + 1)


def camera_index_y(idx: int) -> int:
    return min_camera_y() + idx // (max_camera_x() - min_camera_x() + 1)


def button_camera_x(game_map: GameMap) -> int:
    """Initial camera X guess centred on the emergency button."""
    b = game_map.button
    return max(min_camera_x(), min(max_camera_x(), b.x + b.w // 2 - PLAYER_WORLD_OFF_X))


def button_camera_y(game_map: GameMap) -> int:
    b = game_map.button
    return max(min_camera_y(), min(max_camera_y(), b.y + b.h // 2 - PLAYER_WORLD_OFF_Y))


def camera_x_for_world(x: int) -> int:
    """Camera X that centres one world X on the player."""
    return max(min_camera_x(), min(max_camera_x(), x - PLAYER_WORLD_OFF_X))


def camera_y_for_world(y: int) -> int:
    return max(min_camera_y(), min(max_camera_y(), y - PLAYER_WORLD_OFF_Y))


def in_map(x: int, y: int) -> bool:
    """True when a world pixel is inside the map rectangle."""
    return 0 <= x < MAP_WIDTH and 0 <= y < MAP_HEIGHT


def camera_can_hold_player(cx: int, cy: int) -> bool:
    """True when this camera puts the player's world position inside the map."""
    return in_map(cx + PLAYER_WORLD_OFF_X, cy + PLAYER_WORLD_OFF_Y)


def player_world_x(p: Perception) -> int:
    """Inferred player collision X."""
    return p.camera_x + PLAYER_WORLD_OFF_X


def player_world_y(p: Perception) -> int:
    return p.camera_y + PLAYER_WORLD_OFF_Y


def room_name_at(game_map: GameMap, x: int, y: int) -> str:
    for room in game_map.rooms:
        if room.x <= x < room.x + room.w and room.y <= y < room.y + room.h:
            return room.name
    return "unknown"


def room_name(p: Perception, game_map: GameMap) -> str:
    """Room containing the inferred player position, or ``'unknown'``."""
    if not p.localized:
        return "unknown"
    return room_name_at(game_map, player_world_x(p), player_world_y(p))


def task_center(task: TaskStation) -> tuple[int, int]:
    return (task.cx, task.cy)


def visible_crewmate_world(p: Perception, cx: int, cy: int) -> tuple[int, int]:
    """Convert a screen-coord sprite anchor to world coordinates."""
    return (p.camera_x + cx + SPRITE_DRAW_OFF_X, p.camera_y + cy + SPRITE_DRAW_OFF_Y)


def visible_body_world(p: Perception, bx: int, by: int) -> tuple[int, int]:
    return (p.camera_x + bx + SPRITE_DRAW_OFF_X, p.camera_y + by + SPRITE_DRAW_OFF_Y)


def central_room_center(game_map: GameMap) -> tuple[int, int]:
    """Reference point for the central (button) room."""
    b = game_map.button
    return (b.x + b.w // 2, b.y + b.h // 2)


def central_room_name(game_map: GameMap) -> str:
    cx, cy = central_room_center(game_map)
    return room_name_at(game_map, cx, cy)


def in_central_room(p: Perception, game_map: GameMap) -> bool:
    if not p.localized:
        return False
    central = central_room_name(game_map)
    return central != "unknown" and room_name(p, game_map) == central


def heuristic(ax: int, ay: int, bx: int, by: int) -> int:
    """Manhattan distance — used by A* and by distance-comparison sorting."""
    return abs(ax - bx) + abs(ay - by)


__all__ = [
    "PLAYER_SCREEN_X",
    "PLAYER_SCREEN_Y",
    "PLAYER_WORLD_OFF_X",
    "PLAYER_WORLD_OFF_Y",
    "min_camera_x",
    "max_camera_x",
    "min_camera_y",
    "max_camera_y",
    "camera_index",
    "camera_index_x",
    "camera_index_y",
    "button_camera_x",
    "button_camera_y",
    "camera_x_for_world",
    "camera_y_for_world",
    "in_map",
    "camera_can_hold_player",
    "player_world_x",
    "player_world_y",
    "room_name_at",
    "room_name",
    "task_center",
    "visible_crewmate_world",
    "visible_body_world",
    "central_room_center",
    "central_room_name",
    "in_central_room",
    "heuristic",
]
